core/state/iterator.go

```diff
@@ -0,0 +1,143 @@
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package state
+
+import (
+	"bytes"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/trie"
+)
+
+// NodeIterator is an iterator to traverse the entire state trie post-order,
+// including all of the contract code and contract state tries.
+type NodeIterator struct {
+	state *StateDB // State being iterated
+
+	stateIt *trie.NodeIterator // Primary iterator for the global state trie
+	dataIt  *trie.NodeIterator // Secondary iterator for the data trie of a contract
+
+	accountHash common.Hash // Hash of the node containing the account
+	codeHash    common.Hash // Hash of the contract source code
+	code        []byte      // Source code associated with a contract
+
+	Hash   common.Hash // Hash of the current entry being iterated (nil if not standalone)
+	Entry  interface{} // Current state entry being iterated (internal representation)
+	Parent common.Hash // Hash of the first full ancestor node (nil if current is the root)
+}
+
+// NewNodeIterator creates an post-order state node iterator.
+func NewNodeIterator(state *StateDB) *NodeIterator {
+	return &NodeIterator{
+		state: state,
+	}
+}
+
+// Next moves the iterator to the next node, returning whether there are any
+// further nodes.
+func (it *NodeIterator) Next() bool {
+	it.step()
+	return it.retrieve()
+}
+
+// step moves the iterator to the next entry of the state trie.
+func (it *NodeIterator) step() {
+	// Abort if we reached the end of the iteration
+	if it.state == nil {
+		return
+	}
+	// Initialize the iterator if we've just started
+	if it.stateIt == nil {
+		it.stateIt = trie.NewNodeIterator(it.state.trie.Trie)
+	}
+	// If we had data nodes previously, we surely have at least state nodes
+	if it.dataIt != nil {
+		if cont := it.dataIt.Next(); !cont {
+			it.dataIt = nil
+		}
+		return
+	}
+	// If we had source code previously, discard that
+	if it.code != nil {
+		it.code = nil
+		return
+	}
+	// Step to the next state trie node, terminating if we're out of nodes
+	if cont := it.stateIt.Next(); !cont {
+		it.state, it.stateIt = nil, nil
+		return
+	}
+	// If the state trie node is an internal entry, leave as is
+	if !it.stateIt.Leaf {
+		return
+	}
+	// Otherwise we've reached an account node, initiate data iteration
+	var account struct {
+		Nonce    uint64
+		Balance  *big.Int
+		Root     common.Hash
+		CodeHash []byte
+	}
+	err := rlp.Decode(bytes.NewReader(it.stateIt.LeafBlob), &account)
+	if err != nil {
+		panic(err)
+	}
+	dataTrie, err := trie.New(account.Root, it.state.db)
+	if err != nil {
+		panic(err)
+	}
+	it.dataIt = trie.NewNodeIterator(dataTrie)
+	if !it.dataIt.Next() {
+		it.dataIt = nil
+	}
+	if bytes.Compare(account.CodeHash, emptyCodeHash) != 0 {
```

Comment:
This is really part of the trie iterator PR. I'll leave it for that to include the fix. I've pulled it in here to be able to test the indexing that needs a way to properly iterate over all dependent nodes, but it would be nice to have the iterator merged in independently as it's mostly unrelated code and would just burden the probably way to complex review of this PR.